mlflow/langchain/_langchain_autolog.py

```diff
@@ -0,0 +1,296 @@
+        + f" due to {input_example_exc}."
+        if input_example_exc
+        else ""
+    )
+
+
+def _combine_input_and_output(input, output, session_id, func_name):
+    """
+    Combine input and output into a single dictionary
+    """
+    if func_name == "get_relevant_documents" and output is not None:
+        output = [{"page_content": doc.page_content, "metadata": doc.metadata} for doc in output]
+    if input is None:
+        return {
+            "output": output if isinstance(output, (list, dict)) else [output],
+            "session_id": session_id,
+        }
+    if isinstance(input, (str, dict)):
+        return {"input": [input], "output": [output], "session_id": session_id}
+    if isinstance(input, list) and (
+        all(isinstance(x, str) for x in input) or all(isinstance(x, dict) for x in input)
+    ):
+        if not isinstance(output, list) or len(output) != len(input):
+            raise MlflowException(
+                "Failed to combine input and output data with different lengths "
+                "into a single pandas DataFrame. "
+            )
+        return {"input": input, "output": output, "session_id": session_id}
+    raise MlflowException("Unsupported input type.")
+
+
+def _update_langchain_model_config(model):
+    try:
+        from langchain_core.pydantic_v1 import Extra
+    except ImportError as e:
+        warnings.warn(
+            "MLflow langchain autologging might log model several "
+            "times due to the pydantic.config.Extra import error. "
+            f"Error: {e}"
+        )
+        return False
+    else:
+        if hasattr(model, "__config__"):
+            model.__config__.extra = Extra.allow
+        return True
+
+
+def _inject_mlflow_callback(func_name, mlflow_callback, args, kwargs):
+    if func_name == "invoke":
+        from langchain.schema.runnable.config import RunnableConfig
+
+        in_args = False
+        # `config` is the second positional argument of runnable.invoke function
+        if len(args) >= 2:
+            config = args[1]
+            in_args = True
+        else:
+            config = kwargs.get("config", None)
+        if config is None:
+            callbacks = [mlflow_callback]
+            config = RunnableConfig(callbacks=callbacks)
+        else:
+            callbacks = config.get("callbacks") or []
+            callbacks.append(mlflow_callback)
+            config["callbacks"] = callbacks
+        if in_args:
+            args = (args[0], config) + args[2:]
+        else:
+            kwargs["config"] = config
+        return args, kwargs
+
+    if func_name == "__call__":
+        # `callbacks` is the third positional argument of chain.__call__ function
+        if len(args) >= 3:
+            callbacks = args[2] or []
+            callbacks.append(mlflow_callback)
+            args = args[:2] + (callbacks,) + args[3:]
+        else:
+            callbacks = kwargs.get("callbacks") or []
```

Comment:
```suggestion
            callbacks = kwargs.get("callbacks", [])
```

Reply:
This actually prevents situation where callbacks=None

Reply:
ah makes sense!